File: src/api/instagram.py
"""
Instagram Graph API helper.
Sends text messages back to users via the Instagram Messaging API.
"""
import httpx
from src.config import INSTAGRAM_ACCESS_TOKEN, INSTAGRAM_API_VERSION


async def send_message(recipient_id: str, text: str) -> dict:
    """Send a text message to an Instagram user via the Graph API."""
    url = f"https://graph.instagram.com/{INSTAGRAM_API_VERSION}/me/messages"
    headers = {
        "Authorization": f"Bearer {INSTAGRAM_ACCESS_TOKEN}",
        "Content-Type": "application/json",
    }
    payload = {
        "recipient": {"id": recipient_id},
        "message": {"text": text[:1000]},  # Instagram limit is 1000 chars
    }

    async with httpx.AsyncClient(timeout=10.0) as client:
        response = await client.post(url, json=payload, headers=headers)
        response.raise_for_status()
        return response.json()

File: src/services/instagram.ts
const GRAPH_INSTAGRAM_BASE_URL = "https://graph.instagram.com";
const INSTAGRAM_API_VERSION = "v22.0"; // Ou a versão atualmente utilizada pelo app

function textMessagesUrl(): string {
    return `${GRAPH_INSTAGRAM_BASE_URL}/${INSTAGRAM_API_VERSION}/me/messages`;
}

/**
 * Envia uma mensagem de texto simples pelo Meta Graph API.
 */
export async function sendInstagramMessage(recipientId: string, text: string, token: string): Promise<boolean> {
    const url = textMessagesUrl();
    const safeText = text.substring(0, 1000); // hard limit Meta

    try {
        const res = await fetch(url, {
            method: "POST",
            headers: {
                "Authorization": `Bearer ${token}`,
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                recipient: { id: recipientId },
                message: { text: safeText }
            })
        });

        if (!res.ok) {
            const errorText = await res.text();
            console.error(`[IG API] Falha ao enviar texto: HTTP ${res.status}`, errorText);
            return false;
        }

        console.log(`[IG API] Enviada para ${recipientId.slice(-6)} com sucesso.`);
        return true;
    } catch (error) {
        console.error(`[IG API] Exceção de rede:`, error);
        return false;
    }
}

/**
 * Envia anexo de áudio hospedado no R2 publicamente.
 */
export async function sendInstagramAudio(recipientId: string, audioUrl: string, token: string): Promise<boolean> {
    const url = textMessagesUrl();

    try {
        const res = await fetch(url, {
            method: "POST",
            headers: {
                "Authorization": `Bearer ${token}`,
                "Content-Type": "application/json",
            },
            body: JSON.stringify({
                recipient: { id: recipientId },
                message: {
                    attachment: {
                        type: "audio",
                        payload: { url: audioUrl, is_reusable: false }
                    }
                }
            })
        });

        if (!res.ok) {
            const errorText = await res.text();
            console.error(`[IG API] Falha ao anexar áudio: HTTP ${res.status}`, errorText);
            return false;
        }

        console.log(`[IG API] Áudio enviado com sucesso para ${recipientId.slice(-6)}`);
        return true;
    } catch (error) {
        console.error(`[IG API] Exceção de rede ao enviar áudio:`, error);
        return false;
    }
}
